keras_contrib/applications/resnet.py

```diff
@@ -0,0 +1,282 @@
+    global ROW_AXIS
+    global COL_AXIS
+    global CHANNEL_AXIS
+    if K.image_data_format() == 'channels_last':
+        ROW_AXIS = 1
+        COL_AXIS = 2
+        CHANNEL_AXIS = 3
+    else:
+        CHANNEL_AXIS = 1
+        ROW_AXIS = 2
+        COL_AXIS = 3
+
+
+def _get_block(identifier):
+    if isinstance(identifier, six.string_types):
+        res = globals().get(identifier)
+        if not res:
+            raise ValueError('Invalid {}'.format(identifier))
+        return res
+    return identifier
+
+
+def ResNet(input_shape, classes, block='basic_block', repetitions=[3, 4, 6, 3], initial_filters=64,
+           activation='softmax', include_top=True, input_tensor=None, dropout=None):
+    """Builds a custom ResNet like architecture. Defaults to ResNet50.
+
+    Args:
+        input_shape: The input shape in the form (nb_channels, nb_rows, nb_cols)
+        classes: The number of outputs at final softmax layer
+        block_fn: The block function to use. This is either `'basic_block'` or `'bottleneck'`.
+            The original paper used basic_block for layers < 50
+        repetitions: Number of repetitions of various block units.
+            At each block unit, the number of filters are doubled and the input size is halved
+
+    Returns:
+        The keras `Model`.
+    """
+    _handle_dim_ordering()
+    if len(input_shape) != 3:
+        raise Exception("Input shape should be a tuple (nb_channels, nb_rows, nb_cols)")
+
+    if block == 'basic_block':
+        block_fn = basic_block
+    elif block == 'bottleneck':
+        block_fn = bottleneck
+    else:
+        raise ValueError('Unsupported block type')
+
+    # Permute dimension order if necessary
+    if K.image_data_format() == 'channels_first':
+        input_shape = (input_shape[1], input_shape[2], input_shape[0])
+    # Determine proper input shape
+    input_shape = _obtain_input_shape(input_shape,
+                                      default_size=32,
+                                      min_size=8,
+                                      data_format=K.image_data_format(),
+                                      require_flatten=include_top)
+
+    # Load function from str if needed.
+    block_fn = _get_block(block_fn)
+
+    img_input = Input(shape=input_shape, tensor=input_tensor)
+    conv1 = _conv_bn_relu(filters=initial_filters, kernel_size=(7, 7), strides=(2, 2))(img_input)
+    pool1 = MaxPooling2D(pool_size=(3, 3), strides=(2, 2), padding="same")(conv1)
+
+    block = pool1
+    filters = initial_filters
+    for i, r in enumerate(repetitions):
+        block = _residual_block(block_fn, filters=filters, repetitions=r, is_first_layer=(i == 0),
+                                dropout=dropout)(block)
+        filters *= 2
+
+    # Last activation
+    x = _bn_relu(block)
+
+    # Classifier block
+    block_shape = K.int_shape(block)
+    if include_top:
+        pool2 = AveragePooling2D(pool_size=(block_shape[ROW_AXIS], block_shape[COL_AXIS]),
```

Comment:
Offer different pooling options as a parameter, look at DenseNet

Reply:
hmm, I'll probably have to call them initial_pooling and final_pooling, or top_pooling and bottom_pooling, since there is already a pooling option that differs from densenet...